File: src/org/jgll/util/hashing/StringExternalHasher.java
package org.jgll.util.hashing;

public class StringExternalHasher implements ExternalHasher<String> {
	
	private static final long serialVersionUID = 1L;
	
	private static ExternalHasher<String> instance;
	
	public static ExternalHasher<String> getInstance() {
		if(instance == null) {
			instance = new StringExternalHasher();
		}
		return instance;
	}
	
	@Override
	public int hash(String s, HashFunction f) {
		int[] array = new int[s.length() / 2 + 1];
		
		for (int i = 1; i < s.length(); i += 2) {
			array[i - 1] = s.charAt(i - 1) | (s.charAt(i) << 16);
		}

		return f.hash(array);
	}

	@Override
	public boolean equals(String s1, String s2) {
		return s1.equals(s2);
	}
 
}

File: src-scala/org/jgll/util/hashing/StringExternalHasher.scala
package org.jgll.util.hashing

import org.jgll.util.hashing.hashfunction.HashFunction
import scala.reflect.BeanProperty

object StringExternalHasher {

  @BeanProperty
  lazy val instance = new StringExternalHasher()
}

@SerialVersionUID(1L)
class StringExternalHasher extends ExternalHasher[String] {

  override def hash(s: String, f: HashFunction): Int = {
    val array = Array.ofDim[Int](s.length / 2 + 1)
    var i = 1
    while (i < s.length) {
      array(i - 1) = s.charAt(i - 1) | (s.charAt(i) << 16)
      i += 2
    }
    f.hash(array)
  }

  override def equals(s1: String, s2: String): Boolean = s1 == s2
}
